lib/DataQueue.js

```diff
@@ -0,0 +1,207 @@
+      if (error) {
+        return callback(error, null);
+      }
+      const result = xmlToJson(xmlOutput);
+      const data = result[0].data[0].value;
+      return callback(null, data);
+    });
+  }
+
+   // alias for receive, as the API is "RCV", but jt400 uses the term "read"
+   async read(wait, callback) {
+    return this.receive(wait, callback);
+  }
+
+  //////////////////////////////////////////////////////////////////////////////
+  // Send Data Queue (QSNDDTAQ) API
+  //////////////////////////////////////////////////////////////////////////////
+
+  async send(data, callback) {
+    const program = new ProgramCall('QSNDDTAQ', { lib: 'QSYS' });
+    program.addParam(this.name,    '10A');             // Data queue name
+    program.addParam(this.library, '10A');             // Library name
+    program.addParam(data.length,  '5p0');             // Length of data
+    program.addParam(data,         `${data.length}A`); // Data
+
+    this.connection.add(program)
+    this.connection.run((error, xmlOutput) => {
+      if (error) {
+        return callback(error, null);
+      }
+      const result = xmlToJson(xmlOutput);
+      return callback(null, result);
+    });
+  }
+
+  async write(data, callback) {
+    return this.write(data, callback);
+  }
+
+  //////////////////////////////////////////////////////////////////////////////
+  // Clear Data Queue (QCLRDTAQ) API
+  //////////////////////////////////////////////////////////////////////////////
+
+  async clear() {
+    const program = new ProgramCall('QCLRDTAQ', { lib: 'QSYS' });
+    program.addParam(this.name,    '10A'); // Data queue name
+    program.addParam(this.library, '10A'); // Library name
+
+    this.connection.add(program)
+    this.connection.run((error, xmlOutput) => {
+      if (error) {
+        return callback(error, null);
+      }
+      const result = xmlToJson(xmlOutput);
+      return callback(null, result);
+    });
+  }
+
+  //////////////////////////////////////////////////////////////////////////////
+  // "Private" Functions
+  //////////////////////////////////////////////////////////////////////////////
+
+  
+  async _getDataQueueDescription(callback) {
+    const program = new ProgramCall('QMHQRDQD', { lib: 'QSYS' });
+
+    const receiverVariable = [
+      [0, '10i0'],
+      [0, '10i0'],
+      [0, '10i0']
+    ];
+
+    program.addParam(receiverVariable, { io: 'out', len: 'varLen'}); // Receiver variable
+    program.addParam(0,       '10i0', { setlen: 'varLen'}); // Length of receiver variable
+    program.addParam('RDQD0100',         '8A');   // Format name
+    program.addParam(this.qualifiedName, '20A');  // Qualified data queue name
+
+    this.connection.add(program)
+    this.connection.run((error, xmlOutput) => {
+      if (error) {
+        return callback(error, null);
+      }
+
+      const result = xmlToJson(xmlOutput);
+      this.maxLength = result[0].data[2].value
+      return callback(null, result);
+    });
```

Comment:
Idea here is:

If we create a DataQueue using this API, we know what the max length is. But if we are consuming a DataQueue created elsewhere, we don't know the max size. So when we need to get the length the first time we call an API that requires us to specify the length of the receiver data